solver/src/http_solver.rs

```diff
@@ -58,6 +64,7 @@ impl HttpSolver {
         config: SolverConfig,
         native_token: H160,
         token_info_fetcher: &Arc<dyn TokenInfoFetching>,
+        price_estimator: &Arc<dyn PriceEstimating>,
```

Comment:
Kind of unrelated to this PR, but in this situation, its every so slightly better to move the `Arc<...>` instead of passing a reference.

The reason for this is imagine the code:
```rust
fn foo(val: &Arc<...>) {
  // ...
  val.clone();
  // ...
}

// Some where else:
let val = Arc::new(...);
foo(&val);
```

In this case, we simultaneously have 2 shared pointers to `val` (one in the outer method and one in the inner method). So, reference counting in general is very fast, so the overhead is minimal, but there is one key optimization that you prevent `foo` from doing, and that is using [`Arc::get_mut`](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.get_mut) to get an internal mutable reference, which only works when there is only a single reference to the value.

In general, WRT to ownership, I think its better to ask what the method needs (here an owned `Arc`) instead of a reference and unconditionally `clone`-ing. This gives more control to the caller and can help with overall code optimizations.

This is very small and very nit-picky, and its totally fine as it is here.

Reply:
Thanks for the tip!

I thought it was the opposite. When I see `foo(val)`, without reference, I immediately think that the program will be incrementing the shared ptr count when calling the function, and decrementing when returning, which can be avoided by passing by reference, since that doesn't require cloning the shared_ptr (which changes the ref count under the hoods). 

I suppose this doesn't happen when passing "by move" instead of "by value", which I'm still getting used to :). And if I pass "by move" then I suppose the compiler knows there will be only one pointer to the value (unless it has been cloned before?). so the `get_mut` part makes sense to me. 

So, in general, what is the use case to pass things by const reference instead of moving them?

Reply:
As you said, Rust is by move by default whereas c++ is by copy. In this case the struct is going to store an owned Arc anyway so there is no reason to pass in a reference to this function and then clone internally. By passing in a value you give the caller maximum choice. If they already have an owned Arc that they no longer need they can pass it without copy and if they want to copy they can.

You would pass a reference when you all you need is a reference because then callers don't need to give you an owned value.

Reply:
> I thought it was the opposite. When I see foo(val), without reference, I immediately think that the program will be incrementing the shared ptr count when calling the function, and decrementing when returning,

C++ copy semantics :stuck_out_tongue:. So in Rust, when not using a reference it will `move` instead of `copy`. This means that the reference count will not be incremented, but as you mentioned, it will get decremented once `foo` returns in my example. If the calling context would need to use `val` after calling `foo`, then it would have to:
```rust
let val = Arc::new(...);
foo(val.clone());
// ...
val.do_stuff();
```

This is desired as it allows the caller to control how to best share the resource, i.e. it can decide if it doesn't need it anymore and can move it into `foo`, or if it needs it after `foo`, then

On the other hand, in the code here, the `HttpSolver` needs to unconditionally take ownership of the `Arc`. Passing it by reference here, removes choice from the caller (i.e. the `Arc` will always be cloned, even when it doesn't need to be in the case the caller doesn't need the value any more). Another difference is that since `HttpSolver` is taking ownership of `val`, this means that the shared `Arc` pointer would only be `drop`-ped when the `HttpSolver` gets `drop`-ped.

Going back to my example. If, for example, `foo` only needed a temporary reference to `val`, then you could implement it like:
```rust
fn foo(val: &...) {
  // ...
}

// Some where else:
let val = Arc::new(...);
foo(&val);
```

This allows the caller to provide any reference to the value, regardless if its boxed in a `Box` or a smart pointer like `Arc` or `Rc`. This is possible because [`Arc<T>` implements `Deref<Target = T>`](https://doc.rust-lang.org/std/sync/struct.Arc.html#impl-Deref).


> So, in general, what is the use case to pass things by const reference instead of moving them?

When you only need a reference. In the `HttpSolver::new()` case, it needs an owned value, so the method signature should express that. (Although, to be fair, it is a very nit-picky suggestion, since cloning `Arc` is practically for free).